### ISSUE TITLE
Data type conflict in reduce_over_samples while using torch Tensors

### ISSUE THREAD

**Issue by abmazitov** [lab-cosmo/equistore#247]
There is a data type conflict in `reduce_over_samples`, if it is done on a tensormap with `torch.Tensor` type of data inside.

--- 

In a file `equistore.operations.reduce_over_samples.py` 
In a function `_reduce_over_samples_block()`
Lines 87-89:
```
if reduction == "mean" or reduction == "std" or reduction == "variance":
        bincount = _dispatch.bincount(index)
        values_result = values_result / bincount.reshape(
            (-1,) + (1,) * len(other_shape)
        )
```

`values_result` is a `torch.Tensor`, but `bincount` and `index` are numpy arrays.

The `index` object actually comes from lines 71-73, where datatype management is omitted.

```
 new_samples, index = np.unique(
        samples[:, sample_selected], return_inverse=True, axis=0
    )
``` 

---

This cause a conversion of a torch Tensor to a numpy format and ruins the workflow.

**Comment by abmazitov** [comment]
I would suggest to add a `unique()` function into `_dispatch.py` and maintain a datatype inside

**Comment by DavideTisi** [comment]
mmm i see so index is a numpy array becouse it comes from an operation on a sample which is a numpy array

**Comment by abmazitov** [comment]
> 

Exactly. In this case `samples` has a np.int32 type, and we run `np.unique` on top of it. It means that we have 2 potential sources of unexpected dtype conversion.

**Comment by DavideTisi** [comment]
@Luthaf have any idea how to fix it or we just add some hacks?

**Comment by Luthaf** [comment]
Should be fixed by #250
